tests/cpp/data/test_data.cc

```diff
@@ -127,4 +130,29 @@ TEST(DMatrix, Uri) {
   ASSERT_EQ(dmat->Info().num_row_, kRows);
 }
 
+TEST(DMatrix, DeviceIdx) {
+  size_t constexpr kRows {16};
+  size_t constexpr kCols {8};
+  auto pp_dmat = CreateDMatrix(kRows, kCols, 0);
+  auto& p_dmat = *pp_dmat;
+  ASSERT_EQ(p_dmat->DeviceIdx(), -1);
+  auto& data = (*p_dmat->GetBatches<SparsePage>().begin()).data;
+  data.SetDevice(0);
+  auto& offset = (*p_dmat->GetBatches<SparsePage>().begin()).offset;
+  offset.SetDevice(0);
+
+  // pull data to device
+  data.DeviceSpan();
+  offset.DeviceSpan();
+
+  ASSERT_EQ(p_dmat->DeviceIdx(), 0);
```

Comment:
This cannot work on CPU-only system. Is this test only meant to run in a machine with GPU? https://xgboost-ci.net/blue/organizations/jenkins/xgboost/detail/PR-5066/4/pipeline/107/#step-136-log-648